### PR TITLE
Fix task event handler simul multi warnings

### LINKED ISSUE #2806: Simultaneous message events not handled.
If N message events occur at ~same time - whether by a single invocation of `cylc message` or N separate invocations - only the first gets handled (as in, passed to a configured event handler).

Example:
```
[scheduling]
   [[dependencies]]
      graph = foo
[runtime]
   [[foo]]
      script = """
cylc message -s WARNING -- ${CYLC_SUITE_NAME} ${CYLC_TASK_JOB} "cat"
#sleep 1
cylc message -s WARNING -- ${CYLC_SUITE_NAME} ${CYLC_TASK_JOB} "dog"
#sleep 1
cylc message -s WARNING -- ${CYLC_SUITE_NAME} ${CYLC_TASK_JOB} "fish"
#sleep 1
      """
      [[[events]]]
        handler events = warning
        handlers = echo "HANDLED %(message)s"
```

Run this.  The messages are all logged:
```
$ cylc cat-log foo | grep WARNING
2018-10-17T19:58:31+13 WARNING - [foo.1] -(current:running)> cat at 2018-10-17T19:58:30+13
2018-10-17T19:58:31+13 WARNING - [foo.1] -(current:running)> dog at 2018-10-17T19:58:31+13
2018-10-17T19:58:32+13 WARNING - [foo.1] -(current:running)> fish at 2018-10-17T19:58:31+13
```
But only the first gets handled:
```
$ cylc cat-log -f a foo foo.1 
[jobs-submit ret_code] 0
[jobs-submit out] 2018-10-17T19:58:30+13|1/foo/01|0|18342
2018-10-17T19:58:30+13 [STDOUT] 18342
[(('event-handler-00', 'warning'), 1) cmd] echo "HANDLED cat"
[(('event-handler-00', 'warning'), 1) ret_code] 0
[(('event-handler-00', 'warning'), 1) out] HANDLED cat
```

However, if I uncomment the `# sleep 1`s above, they all get handled:
```
$ cylc cat-log -f a foo foo.1
[jobs-submit ret_code] 0
[jobs-submit out] 2018-10-17T20:06:31+13|1/foo/01|0|19488
2018-10-17T20:06:31+13 [STDOUT] 19488
[(('event-handler-00', 'warning'), 1) cmd] echo "HANDLED cat"
[(('event-handler-00', 'warning'), 1) ret_code] 0
[(('event-handler-00', 'warning'), 1) out] HANDLED cat
[(('event-handler-00', 'warning'), 1) cmd] echo "HANDLED dog"
[(('event-handler-00', 'warning'), 1) ret_code] 0
[(('event-handler-00', 'warning'), 1) out] HANDLED dog
[(('event-handler-00', 'warning'), 1) cmd] echo "HANDLED fish"
[(('event-handler-00', 'warning'), 1) ret_code] 0
[(('event-handler-00', 'warning'), 1) out] HANDLED fish
```
(Interestingly the final `sleep 1` is required too, even with `foo => bar` so that the suite doesn't shut down immediately after foo runs).

### PR DESCRIPTION
**Pull request by matthewrmshin**
Fix #2806.

(Plus new functionality: task output based events).

### CODE REVIEW

File: lib/cylc/task_events_mgr.py [hunk 1]
```diff
@@ -400,11 +401,17 @@ def process_message(
...
+            # this feature can only be used on the deprecated loadleveler
+            # system, we should probably aim to remove support for job vacation
+            # instead. Otherwise, we should have:
+            # self.setup_event_handlers(itask, 'vacated', message)
+        elif completed_trigger:
             # Message of an as-yet unreported custom task output.
             # No state change.
             self.pflag = True
             self.suite_db_mgr.put_update_task_outputs(itask)
+            self.setup_event_handlers(itask, completed_trigger, message)
```
Comment (matthewrmshin) [inline lib/cylc/task_events_mgr.py]: This allows known outputs to be used as events (for generic handler list only).

https://github.com/cylc/cylc/issues/2806#issuecomment-430589075

If desirable, I'll document the new capability.
If not desirable, I'll revert.

Comment (hjoliver) [reply lib/cylc/task_events_mgr.py]: I like it. :+1:

File: lib/cylc/task_events_mgr.py [hunk 2]
```diff
@@ -819,7 +830,11 @@ def _setup_custom_event_handlers(self, itask, event, message):
             retry_delays = [0]
         # There can be multiple custom event handlers
         for i, handler in enumerate(handlers):
-            key1 = ("%s-%02d" % (self.HANDLER_CUSTOM, i), event)
+            if is_generic:
+                key1 = ('%s-%02d' % (self.HANDLER_CUSTOM, i),
+                        '%s-%s' % (event, urllib.quote(message)))
+            else:
+                key1 = ('%s-%02d' % (self.HANDLER_CUSTOM, i), event)
```
Comment (hjoliver) [inline lib/cylc/task_events_mgr.py]: Is this distinction necessary?  If we keyed all events on message the code would be a bit simpler, and it wouldn't actually matter that message isn't needed to uniquely identify non-generic events, would it?

**Comment by hjoliver** [comment]
> alas I don't know much about event handlers in Cylc to give a more useful review here (yet!!!). 

(@matthewrmshin - we're planning to remedy this situation tomorrow :boom:  ... maybe no need to reassign)

File: lib/cylc/task_events_mgr.py (hunk 2 above)
Comment (matthewrmshin) [reply lib/cylc/task_events_mgr.py]: You are right here. I'll investigate the effect of having a single `key1` on our current test battery and on restart, etc.

File: lib/cylc/task_events_mgr.py (hunk 1 above)
Comment (matthewrmshin) [reply lib/cylc/task_events_mgr.py]: In which case, I'll add some basic docs for this. Note this is likely to make #2756 a bit more difficult!

**Comment by kinow** [comment]
@matthewrmshin had a session on event handlers and on this pull request with @hjoliver (thanks!!!). Feeling like I know a bit more about how that works in Cylc and that I can try to help with this review again (bear with me a little longer).

Successfully reproduced the problem on master (left; note that it printed `cat` and `fish`, so the output in master is not deterministic), then used the fix in this pull request and it produced the correct output (right).

ok

But then decided to log fish twice... maybe because I had two files with the same name, or just wanted to raise that message twice. And it was logged once.

other

@matthewrmshin , from what I understood, the solution in this pull request is now using the message as an identifier of the event? If that's correct, then I think this is the desired output.

But would there be cases where users still want to submit the same message multiple times?

Will review the code with more care from home either tonight or during this weekend (a break for when the weather is not so nice, or am too sunburned 🌤 ).

Cheers
Bruno

**Comment by hjoliver** [comment]
@matthewrmshin - actually I think @kinow raises a good point.  I think I can imagine use cases where the same WARNING event, for instance, happens multiple times within the same job, *potentially* with the same message (e.g. "database offline", or whatever).  So maybe we do need to handle every single message as a unique event.

**Comment by hjoliver** [comment]
(@kinow - don't feel obliged to work on the weekend! :sun_with_face:)

**Comment by matthewrmshin** [comment]
In which case, I'll try to use a counter to number the non-unique events.

**Comment by kinow** [comment]
Tested, and works as expected @matthewrmshin ! Thanks for the quick fix! Here's my output

```bash
kinow@ranma:/tmp/foo$ cylc cat-log foo | grep WARNING
2018-10-20T09:16:55+13 WARNING - [foo.1] -(current:running)> cat at 2018-10-20T09:16:54+13
2018-10-20T09:16:55+13 WARNING - [foo.1] -(current:running)> dog at 2018-10-20T09:16:54+13
2018-10-20T09:16:55+13 WARNING - [foo.1] -(current:running)> fish at 2018-10-20T09:16:55+13
2018-10-20T09:16:55+13 WARNING - [foo.1] -(current:running)> fish at 2018-10-20T09:16:55+13
kinow@ranma:/tmp/foo$ cylc cat-log -f a foo foo.1
[jobs-submit ret_code] 0
[jobs-submit out] 2018-10-20T09:15:59+13|1/foo/01|0|6273
2018-10-20T09:15:59+13 [STDOUT] 6273
[(('event-handler-00', 'warning-1'), 1) cmd] echo "HANDLED cat"
[(('event-handler-00', 'warning-1'), 1) ret_code] 0
[(('event-handler-00', 'warning-1'), 1) out] HANDLED cat
[(('event-handler-00', 'warning-3'), 1) cmd] echo "HANDLED fish"
[(('event-handler-00', 'warning-3'), 1) ret_code] 0
[(('event-handler-00', 'warning-3'), 1) out] HANDLED fish
[(('event-handler-00', 'warning-2'), 1) cmd] echo "HANDLED dog"
[(('event-handler-00', 'warning-2'), 1) ret_code] 0
[(('event-handler-00', 'warning-2'), 1) out] HANDLED dog
[jobs-submit ret_code] 0
[jobs-submit out] 2018-10-20T09:16:54+13|1/foo/01|0|6524
2018-10-20T09:16:54+13 [STDOUT] 6524
[(('event-handler-00', 'warning-1'), 1) cmd] echo "HANDLED cat"
[(('event-handler-00', 'warning-1'), 1) ret_code] 0
[(('event-handler-00', 'warning-1'), 1) out] HANDLED cat
[(('event-handler-00', 'warning-3'), 1) cmd] echo "HANDLED fish"
[(('event-handler-00', 'warning-3'), 1) ret_code] 0
[(('event-handler-00', 'warning-3'), 1) out] HANDLED fish
[(('event-handler-00', 'warning-2'), 1) cmd] echo "HANDLED dog"
[(('event-handler-00', 'warning-2'), 1) ret_code] 0
[(('event-handler-00', 'warning-2'), 1) out] HANDLED dog
[(('event-handler-00', 'warning-4'), 1) cmd] echo "HANDLED fish"
[(('event-handler-00', 'warning-4'), 1) ret_code] 0
[(('event-handler-00', 'warning-4'), 1) out] HANDLED fish
```

Minor note, is that in the docs I think we would have to mention that the message would happen twice, but in no guaranteed order. if a user needed messages in order, we would have to include the time in the key? But there is a workaround of asking users to generate a message with the timestamp and sort out the ordering in their side, which could be a messaging system or log handler system.

So LGTM +1 :tada:

**Comment by hjoliver** [comment]
@kinow - regarding ordering, I think this is fine as-is. Normally (in real jobs) there would be a sufficient delay between messages that they will be processed in order by virtue of the fact that they are received one at a time by the suite daemon.  But if messages are emitted near-simultaneously, users should probably not rely on their order.  Plus I can't think of a situation where that would matter.

**Comment by kinow** [comment]
Sure thing @hjoliver , only mentioned because a co-worker in a previous company was telling me about an interesting case he had to use Kafka and keep the order of messages (can't recall what for though). I think it is one of the [guarantees](https://kafka.apache.org/documentation/#intro_guarantees) in Kafka. But agree it may not be worth worrying about it now :+1: what's more important is that this fix appears to be working OK :tada:

File: lib/cylc/task_events_mgr.py (hunk 2 above)
Comment (hjoliver) [reply lib/cylc/task_events_mgr.py]: (holding off on the till you report back on this...)

Comment (matthewrmshin) [reply lib/cylc/task_events_mgr.py]: The latest logic is a bit different. Key points are:
* For unique events, use event as key1. Easy for the eyes (when inspecting the logs) and no change required for tests.
* For non-unique events, use event + count(event) as key1.

Restart looks safe either way.

**Comment by matthewrmshin** [comment]
(Note: event handlers are currently invoked using a parallel pool of subprocesses - so order cannot be guaranteed unless we impose a workflow on them.)

File: lib/cylc/task_events_mgr.py (hunk 2 above)
Comment (hjoliver) [reply lib/cylc/task_events_mgr.py]: OK, seems sensible.

**Comment by hjoliver** [comment]
Merging :fireworks:

File: lib/cylc/task_proxy.py [hunk 3]
```diff
@@ -18,6 +18,7 @@
 
 """Provide a class to represent a task proxy in a running suite."""
 
+from collections import Counter
```
Comment (oliver-sanders) [inline lib/cylc/task_proxy.py]: `collections.Counter` is a Python2.7 feature!

https://docs.python.org/2/library/collections.html

**Comment by hjoliver** [comment]
(Retroactively approving this - I meant to/thought I had before merging ... @oliver-sanders' comment on Counter notwithstanding).

**Comment by hjoliver** [comment]
> collections.Counter is a Python2.7 feature!

Good spotting (very).  Can we just require 2.7 at this point :sheep:-:grin:

**Pull request merged**